pwa/app/lib/matchUtils.ts

```diff
@@ -72,33 +72,121 @@ function matchHasBeenPlayed(match: Match) {
   return match.alliances.red.score !== -1 && match.alliances.blue.score !== -1;
 }
 
-export function calculateTeamRecordFromMatches(
+function getMaybeUnpenalizedScores(m: Match): { blue: number; red: number } {
+  if (m.score_breakdown === null) {
+    return { blue: m.alliances.blue.score, red: m.alliances.red.score };
+  }
+
+  // 2015 used snake_case
+  if (
+    'foul_points' in m.score_breakdown.red &&
+    'foul_points' in m.score_breakdown.blue
+  ) {
+    return {
+      blue: m.alliances.blue.score - (m.score_breakdown.blue.foul_points ?? 0),
+      red: m.alliances.red.score - (m.score_breakdown.red.foul_points ?? 0),
+    };
+  }
+
+  // 2016+ uses camelCase
+  if (
+    'foulPoints' in m.score_breakdown.red &&
+    'foulPoints' in m.score_breakdown.blue
+  ) {
+    return {
+      blue: m.alliances.blue.score - (m.score_breakdown.blue.foulPoints ?? 0),
+      red: m.alliances.red.score - (m.score_breakdown.red.foulPoints ?? 0),
+    };
+  }
+
+  // uhh, this shouldn't happen, but...
+  return { blue: m.alliances.blue.score, red: m.alliances.red.score };
+}
+
+export function calculateTeamRecordsFromMatches(
   teamKey: string,
   matches: Match[],
-): WltRecord {
+): {
+  quals: WltRecord;
+  playoff: WltRecord;
+} {
   const won = matches.filter(
     (m) =>
       (m.winning_alliance === 'red' &&
         m.alliances.red.team_keys.includes(teamKey)) ||
       (m.winning_alliance === 'blue' &&
         m.alliances.blue.team_keys.includes(teamKey)),
-  ).length;
+  );
 
   const lost = matches.filter(
     (m) =>
       (m.winning_alliance === 'red' &&
         m.alliances.blue.team_keys.includes(teamKey)) ||
       (m.winning_alliance === 'blue' &&
         m.alliances.red.team_keys.includes(teamKey)),
-  ).length;
+  );
 
   const tied = matches.filter(
     (m) =>
       (m.alliances.blue.team_keys.includes(teamKey) ||
         m.alliances.red.team_keys.includes(teamKey)) &&
       m.winning_alliance === '' &&
       matchHasBeenPlayed(m),
-  ).length;
+  );
+
+  const quals: WltRecord = {
+    wins: won.filter((m) => m.comp_level === 'qm').length,
+    losses: lost.filter((m) => m.comp_level === 'qm').length,
+    ties: tied.filter((m) => m.comp_level === 'qm').length,
+  };
+
+  const playoff: WltRecord = {
+    wins: won.filter((m) => m.comp_level !== 'qm').length,
+    losses: lost.filter((m) => m.comp_level !== 'qm').length,
+    ties: tied.filter((m) => m.comp_level !== 'qm').length,
+  };
+
+  return { quals, playoff };
+}
+
+export function getTeamsUnpenalizedHighScore(
+  teamKey: string,
+  matches: Match[],
+):
+  | {
+      match: Match;
+      score: number;
+      alliance: MatchAlliance;
+    }
+  | undefined {
+  if (matches.length === 0) {
+    return undefined;
+  }
+
+  const highScoreMatch = matches
```

Comment:
What happens here if teamKey doesn't exist in matches? Does `[0]` break? maybe it makes sense to filter by teamKey first, check for length, and then map for the high score?